Add TorchBindOpOverload

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123370
* #123367
* #123360
* __->__ #123358

We create a new subclass of OpOverload called TorchBindOpOverload, which refers to those custom ops whose arguments consist of torch.ScriptObject.

This PR just adds this new OpOverload and have some tests to make sure they can be created. We'll have concrete implementations in later of the stack.

ydwu4 committed:
Add TorchBindOpOverload

[ghstack-poisoned]

zou3519 commented on torch/_ops.py:
```diff
@@ -782,6 +782,20 @@ def tags(self):
+# schema consists of torch.ScriptObject (i.e. custom class) input.
+# TorchBindOpOverload will skip C++ dispatcher and purely dispatched in python
+# when its inputs contain FakeScriptObject.
+class TorchBindOpOverload(OpOverload):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._is_torchbind_op = True
```
Why do you need a separate TorchBindOpOverload? `_is_torchbind_op` on an OpOverload isn't sufficient?

ydwu4 replied:
We also need to add a new `__call__` and `__init__` logic in OpOverload later in the stack so I think it might be cleaner to just create a new subclass and put all its logic together. The alternative is to write a few `if _is_torchbind_op` checks, which seems a bit more messy.

ydwu4 committed:
Update on "Add TorchBindOpOverload"


We create a new subclass of OpOverload called TorchBindOpOverload, which refers to those custom ops whose arguments consist of torch.ScriptObject.

This PR just adds this new OpOverload and have some tests to make sure they can be created. We'll have concrete implementations in later of the stack.


[ghstack-poisoned]

zou3519 commented on torch/_ops.py:
```diff
@@ -782,6 +782,20 @@ def tags(self):
+class TorchBindOpOverload(OpOverload):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._is_torchbind_op = True
```
Why do we need a `self._is_torchbind_op`? It seems like one can just do isinstance(self, torch._ops.TorchBindOpOverload)

zou3519 commented on torch/_ops.py:
```diff
@@ -845,8 +868,18 @@ def __getattr__(self, key):
-            schema = torch._C._get_schema(self._qualified_op_name, use_key)
-            overload = OpOverload(self, op_, op_dk_, schema, tags)
+
+            if use_key not in self._schemas:
+                # Raise Runtime error is to be consistant with the error raised by torch._C._get_schema
+                raise RuntimeError(f"Found no matching schema for {use_key}")
+
+            schema = self._schemas[use_key]
+
+            overload = (
+                OpOverload(self, op_, op_dk_, schema, tags)
+                if not self._overload_has_script_obj_arg[use_key]
+                else TorchBindOpOverload(self, op_, op_dk_, schema, tags)
+            )
```
We don't need `self._schemas` and `self._overload_has_script_obj_arg`, right? In general we want to avoid the redundancy. If the motivation behind this is caching the attributes for performance: we don't need to do this: `__getattr__` only gets called if the OpOverload can't be found

zou3519 replied:
So the logic here should just be:
```py
schema = torch._C._get_schema(self._qualified_op_name, use_key)
            overload = OpOverload(self, op_, op_dk_, schema, 
if has_script_object(schema):
    return TorchBindOpOverload(...)
else:
    return OpOverload
```

zou3519 reviewed (commented):
Subclass OpOverload seems good. I think the new fields on OpOverload/TorchBindOpOverload are redundant; let's remove them.
